### DISCUSSION TITLE
AutomaticRecovery limitations

### DISCUSSION THREAD

**Discussion by guru1306** [rabbitmq/rabbitmq-java-client#704] [Q&A]
Hi All

Automatic recovery has limitations as per official [blog](https://www.rabbitmq.com/api-guide.html#automatic-recovery-limitations). How do I Identify channel-level exceptions in code? 
```
Connection recovery will not kick in when a channel is closed due to a channel-level exception. Such exceptions often indicate application-level issues. The library cannot make an informed decision about when that's the case.

Closed channels won't be recovered even after connection recovery kicks in. This includes both explicitly closed channels and the channel-level exception case above
```

Can I check if it's not a hard error in the shutdownSignalException and then attempt to create a new connection as mentioned [here](https://www.rabbitmq.com/api-guide.html#shutdown-cause)?

THanks
Guru

**Comment by acogoluegnes** [comment]
You can customize the condition that triggers connection recovery [1]. The default is [2] [3]. I don't know exactly what you're doing, but you should be careful not to trigger connection recovery to easily by changing the condition.

[1] https://rabbitmq.github.io/rabbitmq-java-client/api/current/com/rabbitmq/client/ConnectionFactory.html#setConnectionRecoveryTriggeringCondition(java.util.function.Predicate)
[2] https://rabbitmq.github.io/rabbitmq-java-client/api/current/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.html#DEFAULT_CONNECTION_RECOVERY_TRIGGERING_CONDITION
[3] https://github.com/rabbitmq/rabbitmq-java-client/blob/12f9f75bdf973ed74082df5a0fc5d664610c1595/src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.java#L65-L66

**Comment by guru1306** [comment]
Thanks. I am adding this check in the error as soft errors are channel errors. I am closing the connection and channel. The application will recover from another code path. Is that okay?

![image](https://user-images.githubusercontent.com/12997511/128474342-5b0d9e9b-8c61-4e2c-968a-6d88897321d4.png)

**Reply by acogoluegnes** [reply]
I'm not sure what you're trying to achieve. You'd like to make sure the channel is usable? If so you may want to catch application-level exceptions on channels and also cast connection instances to `Recoverable` and plug a `RecoveryListener` on them to know when the connection is actually recovering and use it accordingly.

**Reply by guru1306** [reply]
Since the auto-recovery doesn't work in case of channel-level errors, I am trying to handle that case by closing the connection and open it by explicitly creating another connection before publishing the message

I did some testing and closing the connection is not required.  Will the rabbitmq client recover for all channel-level errors? The RMQ [blog  ](https://www.rabbitmq.com/api-guide.html#automatic-recovery-limitations) says that it doesn't handle certain channel-level errors. Is there a condition to check if there is a non-recoverable channel level error.  This will help the application to create a new channel and update the existing subscriptions?
